Test the cargo args generated by bootstrap.py

jyn514 opened the pull request:
I recommend reviewing this commit-by-commit using the instructions in https://rustc-dev-guide.rust-lang.org/git.html#moving-large-sections-of-code.

- Test cargo arguments passed by bootstrap.py

  This moves a lot of code around, but the logic itself is not too terribly complicated.

    - Move almost all logic in `def bootstrap` to the `RustBuild` class, to avoid mixing setting configuration with running commands
    - Update various doctests to the new (more complete) RustBuild config. In particular, don't pretend that `bin_root` supports `build` being unset.
    - Change `parse_args` not to use a global, to allow testing it
    - Set BUILD_DIR appropriately so bootstrap.py doesn't panic because cargo isn't found

- Allow passing arguments to `bootstrap_test.py`

    Previous, it used the built-in test runner, which doesn't support options unless they're manually passed in the script.

- Fix progress messages for configure in bootstrap_test.py
    
    Before it would unconditionally print `configure-args = []`.

r? @albertlarsan68 cc https://github.com/rust-lang/rust/pull/112089 https://github.com/rust-lang/rust/pull/111979#issuecomment-1568525699

jyn514 committed:
Fix progress messages for configure in bootstrap_test.py

Before it would unconditionally print `configure-args = []`.

jyn514 committed:
Allow passing arguments to `bootstrap_test.py`

Previous, it used the built-in test runner, which doesn't support options unless they're manually passed in the script.

jyn514 committed:
Test cargo arguments passed by bootstrap.py

This moves a lot of code around, but the logic itself is not too terribly complicated.

- Move almost all logic in `def bootstrap` to the `RustBuild` class, to avoid mixing setting configuration with running commands
- Update various doctests to the new (more complete) RustBuild config. In particular, don't pretend that `bin_root` supports `build` being unset.
- Change `parse_args` not to use a global, to allow testing it
- Set BUILD_DIR appropriately so bootstrap.py doesn't panic because cargo isn't found

jyn514 commented:
oh i want to do one more refactor related to `verbose_count` - let me know if i should put it in a new pr or not

jyn514 committed:
Test color/verbose/warnings properly

These weren't being passed in to bootstrap consistently before; in particular `serialize_and_parse` forgot to pass them in.

albertlarsan68 commented on src/bootstrap/bootstrap.py:
```diff
@@ -458,23 +458,52 @@ def unpack_component(download_info):
-        self.download_url = ''
+class FakeArgs:
+    """Used for unit tests to avoid updating all call sites"""
+    def __init__(self) -> None:
```
Is this syntax supported by python 2.7?

jyn514 replied:
whoops, my editor generated this. will fix it

albertlarsan68 commented on src/bootstrap/test.rs:
```diff
@@ -2529,7 +2529,13 @@ impl Step for Bootstrap {
         let mut check_bootstrap = Command::new(&builder.python());
-        check_bootstrap.arg("bootstrap_test.py").current_dir(builder.src.join("src/bootstrap/"));
+        check_bootstrap
+            .arg("-m")
+            .arg("unittest")
+            .arg("bootstrap_test.py")
```
```suggestion
            .args(["-m", "unittest", "bootstrap_test.py"])
```

Just personal preference, maybe the convention is the other way.

albertlarsan68 reviewed (commented):
Looks mostly good to me, r=me with questions answered.

jyn514 commented:
@bors r=albertlarsan68

bors commented:
:pushpin: Commit c2723c70b1fd31de5a57558ad1e373d0e0f36052 has been approved by `albertlarsan68`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=albertlarsan68 c2723c70b1fd31de5a57558ad1e373d0e0f36052 -->
<!-- homu: {"type":"Approved","sha":"c2723c70b1fd31de5a57558ad1e373d0e0f36052","approver":"albertlarsan68","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit c2723c70b1fd31de5a57558ad1e373d0e0f36052 with merge bdae42044ea1244db04c548301491ce5f1919556...
<!-- homu: {"type":"BuildStarted","head_sha":"c2723c70b1fd31de5a57558ad1e373d0e0f36052","merge_sha":"bdae42044ea1244db04c548301491ce5f1919556"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/5207197271/jobs/9394510936) [(plain)](https://github.com/rust-lang-ci/rust/commit/bdae42044ea1244db04c548301491ce5f1919556/checks/14094604575/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    found:    334d016f755cd6dc58c53a86e183882f8ec14f52fb05345887c8a5edd42c87b7
    expected: 64ec88ca00b268e5ba1a35678a1b5316d212f4f366b2477232534a8aeca37f3c
..
======================================================================
ERROR: test_cargoflags (bootstrap_test.BuildBootstrap.test_cargoflags)
Traceback (most recent call last):
  File "/checkout/src/bootstrap/bootstrap_test.py", line 138, in test_cargoflags
  File "/checkout/src/bootstrap/bootstrap_test.py", line 138, in test_cargoflags
    args, _ = self.build_args(env={"CARGOFLAGS": "--timings"})
  File "/checkout/src/bootstrap/bootstrap_test.py", line 133, in build_args
    build = serialize_and_parse(configure_args, parsed)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/checkout/src/bootstrap/bootstrap_test.py", line 29, in serialize_and_parse
  File "/checkout/src/bootstrap/bootstrap_test.py", line 29, in serialize_and_parse
    tomllib.loads(build.config_toml)
  File "/usr/lib/python3.11/tomllib/_parser.py", line 113, in loads
    pos, header = create_dict_rule(src, pos, out)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/tomllib/_parser.py", line 290, in create_dict_rule
    raise suffixed_err(src, pos, f"Cannot declare {key} twice")
tomllib.TOMLDecodeError: Cannot declare ('build',) twice (at line 32, column 7)
======================================================================
======================================================================
ERROR: test_warnings (bootstrap_test.BuildBootstrap.test_warnings)
Traceback (most recent call last):
  File "/checkout/src/bootstrap/bootstrap_test.py", line 147, in test_warnings
  File "/checkout/src/bootstrap/bootstrap_test.py", line 147, in test_warnings
    _, env = self.build_args(configure_args, args=["--warnings=warn"])
  File "/checkout/src/bootstrap/bootstrap_test.py", line 133, in build_args
    build = serialize_and_parse(configure_args, parsed)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/checkout/src/bootstrap/bootstrap_test.py", line 29, in serialize_and_parse
  File "/checkout/src/bootstrap/bootstrap_test.py", line 29, in serialize_and_parse
    tomllib.loads(build.config_toml)
  File "/usr/lib/python3.11/tomllib/_parser.py", line 113, in loads
    pos, header = create_dict_rule(src, pos, out)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/tomllib/_parser.py", line 290, in create_dict_rule
    raise suffixed_err(src, pos, f"Cannot declare {key} twice")
tomllib.TOMLDecodeError: Cannot declare ('build',) twice (at line 35, column 7)
======================================================================
ERROR: test_no_args (bootstrap_test.GenerateAndParseConfig.test_no_args)
----------------------------------------------------------------------
Traceback (most recent call last):
Traceback (most recent call last):
  File "/checkout/src/bootstrap/bootstrap_test.py", line 99, in test_no_args
    build = serialize_and_parse([])
            ^^^^^^^^^^^^^^^^^^^^^^^
  File "/checkout/src/bootstrap/bootstrap_test.py", line 29, in serialize_and_parse
    tomllib.loads(build.config_toml)
  File "/usr/lib/python3.11/tomllib/_parser.py", line 113, in loads
    pos, header = create_dict_rule(src, pos, out)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/tomllib/_parser.py", line 290, in create_dict_rule
    raise suffixed_err(src, pos, f"Cannot declare {key} twice")
tomllib.TOMLDecodeError: Cannot declare ('build',) twice (at line 32, column 7)
======================================================================
ERROR: test_set_codegen_backends (bootstrap_test.GenerateAndParseConfig.test_set_codegen_backends)
----------------------------------------------------------------------
Traceback (most recent call last):
Traceback (most recent call last):
  File "/checkout/src/bootstrap/bootstrap_test.py", line 117, in test_set_codegen_backends
    build = serialize_and_parse(["--set", "rust.codegen-backends=cranelift"])
  File "/checkout/src/bootstrap/bootstrap_test.py", line 29, in serialize_and_parse
  File "/checkout/src/bootstrap/bootstrap_test.py", line 29, in serialize_and_parse
    tomllib.loads(build.config_toml)
  File "/usr/lib/python3.11/tomllib/_parser.py", line 113, in loads
    pos, header = create_dict_rule(src, pos, out)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/tomllib/_parser.py", line 290, in create_dict_rule
    raise suffixed_err(src, pos, f"Cannot declare {key} twice")
tomllib.TOMLDecodeError: Cannot declare ('build',) twice (at line 39, column 7)
======================================================================
ERROR: test_set_section (bootstrap_test.GenerateAndParseConfig.test_set_section)
----------------------------------------------------------------------
Traceback (most recent call last):
Traceback (most recent call last):
  File "/checkout/src/bootstrap/bootstrap_test.py", line 105, in test_set_section
    build = serialize_and_parse(["--set", "llvm.download-ci-llvm"])
  File "/checkout/src/bootstrap/bootstrap_test.py", line 29, in serialize_and_parse
  File "/checkout/src/bootstrap/bootstrap_test.py", line 29, in serialize_and_parse
    tomllib.loads(build.config_toml)
  File "/usr/lib/python3.11/tomllib/_parser.py", line 113, in loads
    pos, header = create_dict_rule(src, pos, out)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/tomllib/_parser.py", line 290, in create_dict_rule
    raise suffixed_err(src, pos, f"Cannot declare {key} twice")
tomllib.TOMLDecodeError: Cannot declare ('build',) twice (at line 46, column 7)
======================================================================
ERROR: test_set_target (bootstrap_test.GenerateAndParseConfig.test_set_target)
----------------------------------------------------------------------
Traceback (most recent call last):
Traceback (most recent call last):
  File "/checkout/src/bootstrap/bootstrap_test.py", line 109, in test_set_target
    build = serialize_and_parse(["--set", "target.x86_64-unknown-linux-gnu.cc=gcc"])
  File "/checkout/src/bootstrap/bootstrap_test.py", line 29, in serialize_and_parse
  File "/checkout/src/bootstrap/bootstrap_test.py", line 29, in serialize_and_parse
    tomllib.loads(build.config_toml)
  File "/usr/lib/python3.11/tomllib/_parser.py", line 113, in loads
    pos, header = create_dict_rule(src, pos, out)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/tomllib/_parser.py", line 290, in create_dict_rule
    raise suffixed_err(src, pos, f"Cannot declare {key} twice")
tomllib.TOMLDecodeError: Cannot declare ('build',) twice (at line 38, column 7)
======================================================================
ERROR: test_set_top_level (bootstrap_test.GenerateAndParseConfig.test_set_top_level)
----------------------------------------------------------------------
Traceback (most recent call last):
Traceback (most recent call last):
  File "/checkout/src/bootstrap/bootstrap_test.py", line 113, in test_set_top_level
    build = serialize_and_parse(["--set", "profile=compiler"])
  File "/checkout/src/bootstrap/bootstrap_test.py", line 29, in serialize_and_parse
  File "/checkout/src/bootstrap/bootstrap_test.py", line 29, in serialize_and_parse
    tomllib.loads(build.config_toml)
  File "/usr/lib/python3.11/tomllib/_parser.py", line 113, in loads
    pos, header = create_dict_rule(src, pos, out)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/tomllib/_parser.py", line 290, in create_dict_rule
    raise suffixed_err(src, pos, f"Cannot declare {key} twice")
tomllib.TOMLDecodeError: Cannot declare ('build',) twice (at line 32, column 7)
----------------------------------------------------------------------
Ran 12 tests in 0.040s

FAILED (errors=7)
FAILED (errors=7)
configure: build.configure-args := []
configure: profile              := user
configure: build.configure-args := ['--set', 'rust.deny-warnings=false']
configure: rust.deny-warnings   := False
configure: profile              := user
configure: build.configure-args := []
configure: profile              := user
configure: build.configure-args := ['--set', 'rust.codegen-backends=cranelift']
configure: profile              := user
configure: build.configure-args := ['--set', 'llvm.download-ci-llvm']
configure: llvm.download-ci-llvm := True
configure: profile              := user
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5207197271/jobs/9395044066)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/5207197271/jobs/9395044066","builder_name":"checks-actions"} -->

albertlarsan68 commented:
It seems like the hack to allow for two levels of defaults is not working anymore.

jyn514 committed:
Don't test the profile override hack

It generates invalid TOML. I want to get rid of it eventually, but this avoids the issue in the meantime.

albertlarsan68 commented on src/bootstrap/bootstrap.py:
```diff
@@ -795,8 +785,13 @@ def get_toml(self, key, section=None):
         'true'
         """
 
+        if type(self) == str:
+            config_toml = self
+        else:
+            config_toml = self.config_toml
```
I don't really understand how `self` can be something else than `RustBuild`, especially `str`.

jyn514 replied:
yeah, this was kinda hacky. if you look at the callsite, it's calling `RustBuild.get_toml` directly instead of `build.get_toml`: https://github.com/rust-lang/rust/blob/db8f6a095484a4226e65d12b52656f5529795d01/src/bootstrap/bootstrap.py#L1052 i did that to avoid having to add a free-standing get_toml function in addition to the class function, but i don't mind adding the free function if you prefer it.

albertlarsan68 replied:
If this would be documented, it will also be fine.

jyn514 replied:
don't want to write documentation for this, i'll add the free function

bors commented:
:umbrella: The latest upstream changes (presumably #112974) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jyn514 commented:
@bors r=albertlarsan68

bors commented:
:pushpin: Commit 24e67d51a0fbf20910e19045e038fe646e5b0910 has been approved by `albertlarsan68`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=albertlarsan68 24e67d51a0fbf20910e19045e038fe646e5b0910 -->
<!-- homu: {"type":"Approved","sha":"24e67d51a0fbf20910e19045e038fe646e5b0910","approver":"albertlarsan68","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 24e67d51a0fbf20910e19045e038fe646e5b0910 with merge b716bc479aaeb4143d9f05a9390fb9d659d57525...
<!-- homu: {"type":"BuildStarted","head_sha":"24e67d51a0fbf20910e19045e038fe646e5b0910","merge_sha":"b716bc479aaeb4143d9f05a9390fb9d659d57525"} -->

rust-log-analyzer commented:
The job **`i686-msvc`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/5366993848/jobs/9736792611) [(plain)](https://github.com/rust-lang-ci/rust/commit/b716bc479aaeb4143d9f05a9390fb9d659d57525/checks/14528279427/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    found:    334d016f755cd6dc58c53a86e183882f8ec14f52fb05345887c8a5edd42c87b7
    expected: 64ec88ca00b268e5ba1a35678a1b5316d212f4f366b2477232534a8aeca37f3c
..
======================================================================
ERROR: test_cargoflags (bootstrap_test.BuildBootstrap.test_cargoflags)
Traceback (most recent call last):
  File "C:\a\rust\rust\src\bootstrap\bootstrap_test.py", line 138, in test_cargoflags
  File "C:\a\rust\rust\src\bootstrap\bootstrap_test.py", line 138, in test_cargoflags
    args, _ = self.build_args(env={"CARGOFLAGS": "--timings"})
  File "C:\a\rust\rust\src\bootstrap\bootstrap_test.py", line 135, in build_args
configure: build.configure-args := []
configure: profile              := user
configure: build.configure-args := ['--set', 'rust.deny-warnings=false']
configure: build.configure-args := ['--set', 'rust.deny-warnings=false']
configure: rust.deny-warnings   := False
configure: profile              := user
configure: build.configure-args := []
    return build.build_bootstrap_cmd(env), env
  File "C:\a\rust\rust\src\bootstrap\bootstrap.py", line 943, in build_bootstrap_cmd
  File "C:\a\rust\rust\src\bootstrap\bootstrap.py", line 943, in build_bootstrap_cmd
    raise Exception("no cargo executable found at `{}`".format(
Exception: no cargo executable found at `C:\a\rust\rust\build\x86_64-pc-windows-gnu\stage0\bin\cargo.exe`
======================================================================
======================================================================
ERROR: test_warnings (bootstrap_test.BuildBootstrap.test_warnings)
Traceback (most recent call last):
  File "C:\a\rust\rust\src\bootstrap\bootstrap_test.py", line 147, in test_warnings
  File "C:\a\rust\rust\src\bootstrap\bootstrap_test.py", line 147, in test_warnings
    _, env = self.build_args(configure_args, args=["--warnings=warn"])
  File "C:\a\rust\rust\src\bootstrap\bootstrap_test.py", line 135, in build_args
  File "C:\a\rust\rust\src\bootstrap\bootstrap_test.py", line 135, in build_args
    return build.build_bootstrap_cmd(env), env
  File "C:\a\rust\rust\src\bootstrap\bootstrap.py", line 943, in build_bootstrap_cmd
  File "C:\a\rust\rust\src\bootstrap\bootstrap.py", line 943, in build_bootstrap_cmd
    raise Exception("no cargo executable found at `{}`".format(
Exception: no cargo executable found at `C:\a\rust\rust\build\x86_64-pc-windows-gnu\stage0\bin\cargo.exe`
----------------------------------------------------------------------
Ran 12 tests in 0.621s

FAILED (errors=2)
---
configure: profile              := user
configure: build.configure-args := ['--set', 'profile=compiler']
configure: profile              := compiler
Build completed unsuccessfully in 1:12:03
make: *** [Makefile:66: ci-msvc-ps1] Error 1
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5366993848/jobs/9737317961)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/5366993848/jobs/9737317961","builder_name":"checks-actions"} -->

jyn514 committed:
fix some bugs

- fix tests when `--build` is set
- don't leak `config.example.toml` fd
- don't crash if `config.toml` doesn't exist yet

albertlarsan68 commented:
Thanks for the PR!
@bors r+

bors commented:
:pushpin: Commit 1e7f03718b3e1c19e59afff518d80d3717b785b9 has been approved by `albertlarsan68`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=albertlarsan68 1e7f03718b3e1c19e59afff518d80d3717b785b9 -->
<!-- homu: {"type":"Approved","sha":"1e7f03718b3e1c19e59afff518d80d3717b785b9","approver":"albertlarsan68","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

clubby789 commented on src/bootstrap/test.rs:
```diff
@@ -2664,7 +2664,12 @@ impl Step for Bootstrap {
+            .env("BUILD_DIR", &builder.out)
+            .env("BUILD_PLATFORM", &builder.build.build.triple)
+            .current_dir(builder.src.join("src/bootstrap/"))
+            .args(builder.config.test_args());
```
This means that `x test src/bootstrap -- <rust_test_name>` no longer works, as `unittest` throws an exception when the test name doesn't exist as a module and stops the testing

jyn514 replied:
i'm going to revert this for now - people can use `python -m pytest` manually if they want to pass test args.
